File: julia/src/sorting.jl
"""
    Sorting

Module containing various sorting algorithms implemented in Julia.
"""
module Sorting

export quicksort, quicksort!, mergesort, mergesort!,
       heapsort!, insertionsort!, selectionsort!, bubblesort!,
       shellsort!, countingsort, radixsort, bucketsort,
       timsort!, introsort!, quickselect

"""
    quicksort(arr::Vector{T}) where T

Return a sorted copy of the array using quicksort algorithm.

# Time Complexity
- Best/Average: O(n log n)
- Worst: O(n²)
- Space: O(log n)
"""
function quicksort(arr::Vector{T}) where T
    result = copy(arr)
    quicksort!(result)
    return result
end

"""
    quicksort!(arr::Vector{T}, lo::Int=1, hi::Int=length(arr)) where T

Sort array in-place using quicksort algorithm.
"""
function quicksort!(arr::Vector{T}, lo::Int=1, hi::Int=length(arr)) where T
    if lo < hi
        p = partition!(arr, lo, hi)
        quicksort!(arr, lo, p - 1)
        quicksort!(arr, p + 1, hi)
    end
    return arr
end

function partition!(arr::Vector{T}, lo::Int, hi::Int) where T
    pivot = arr[hi]
    i = lo - 1

    for j in lo:hi-1
        if arr[j] <= pivot
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
        end
    end

    arr[i + 1], arr[hi] = arr[hi], arr[i + 1]
    return i + 1
end

"""
    mergesort(arr::Vector{T}) where T

Return a sorted copy of the array using mergesort algorithm.

# Time Complexity
- All cases: O(n log n)
- Space: O(n)
"""
function mergesort(arr::Vector{T}) where T
    if length(arr) <= 1
        return copy(arr)
    end

    mid = length(arr) ÷ 2
    left = mergesort(arr[1:mid])
    right = mergesort(arr[mid+1:end])

    return merge(left, right)
end

"""
    mergesort!(arr::Vector{T}) where T

Sort array in-place using mergesort algorithm (with auxiliary space).
"""
function mergesort!(arr::Vector{T}) where T
    n = length(arr)
    aux = similar(arr)
    mergesort_helper!(arr, aux, 1, n)
    return arr
end

function mergesort_helper!(arr::Vector{T}, aux::Vector{T}, lo::Int, hi::Int) where T
    if lo >= hi
        return
    end

    mid = lo + (hi - lo) ÷ 2
    mergesort_helper!(arr, aux, lo, mid)
    mergesort_helper!(arr, aux, mid + 1, hi)
    merge!(arr, aux, lo, mid, hi)
end

function merge!(arr::Vector{T}, aux::Vector{T}, lo::Int, mid::Int, hi::Int) where T
    # Copy to auxiliary array
    for k in lo:hi
        aux[k] = arr[k]
    end

    i = lo
    j = mid + 1

    for k in lo:hi
        if i > mid
            arr[k] = aux[j]
            j += 1
        elseif j > hi
            arr[k] = aux[i]
            i += 1
        elseif aux[j] < aux[i]
            arr[k] = aux[j]
            j += 1
        else
            arr[k] = aux[i]
            i += 1
        end
    end
end

function merge(left::Vector{T}, right::Vector{T}) where T
    result = T[]
    i = j = 1

    while i <= length(left) && j <= length(right)
        if left[i] <= right[j]
            push!(result, left[i])
            i += 1
        else
            push!(result, right[j])
            j += 1
        end
    end

    append!(result, left[i:end])
    append!(result, right[j:end])

    return result
end

"""
    heapsort!(arr::Vector{T}) where T

Sort array in-place using heapsort algorithm.

# Time Complexity
- All cases: O(n log n)
- Space: O(1)
"""
function heapsort!(arr::Vector{T}) where T
    n = length(arr)

    # Build max heap
    for i in n÷2:-1:1
        heapify!(arr, n, i)
    end

    # Extract elements from heap
    for i in n:-1:2
        arr[1], arr[i] = arr[i], arr[1]
        heapify!(arr, i - 1, 1)
    end

    return arr
end

function heapify!(arr::Vector{T}, n::Int, i::Int) where T
    largest = i
    left = 2 * i
    right = 2 * i + 1

    if left <= n && arr[left] > arr[largest]
        largest = left
    end

    if right <= n && arr[right] > arr[largest]
        largest = right
    end

    if largest != i
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify!(arr, n, largest)
    end
end

"""
    insertionsort!(arr::Vector{T}, lo::Int=1, hi::Int=length(arr)) where T

Sort array in-place using insertion sort algorithm.

# Time Complexity
- Best: O(n)
- Average/Worst: O(n²)
- Space: O(1)
"""
function insertionsort!(arr::Vector{T}, lo::Int=1, hi::Int=length(arr)) where T
    for i in lo+1:hi
        key = arr[i]
        j = i - 1

        while j >= lo && arr[j] > key
            arr[j + 1] = arr[j]
            j -= 1
        end

        arr[j + 1] = key
    end

    return arr
end

"""
    selectionsort!(arr::Vector{T}) where T

Sort array in-place using selection sort algorithm.

# Time Complexity
- All cases: O(n²)
- Space: O(1)
"""
function selectionsort!(arr::Vector{T}) where T
    n = length(arr)

    for i in 1:n-1
        min_idx = i

        for j in i+1:n
            if arr[j] < arr[min_idx]
                min_idx = j
            end
        end

        if min_idx != i
            arr[i], arr[min_idx] = arr[min_idx], arr[i]
        end
    end

    return arr
end

"""
    bubblesort!(arr::Vector{T}) where T

Sort array in-place using bubble sort algorithm.

# Time Complexity
- Best: O(n)
- Average/Worst: O(n²)
- Space: O(1)
"""
function bubblesort!(arr::Vector{T}) where T
    n = length(arr)

    for i in 1:n-1
        swapped = false

        for j in 1:n-i
            if arr[j] > arr[j + 1]
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = true
            end
        end

        if !swapped
            break
        end
    end

    return arr
end

"""
    shellsort!(arr::Vector{T}) where T

Sort array in-place using shell sort algorithm.

# Time Complexity
- Best: O(n log n)
- Average: O(n^1.5)
- Worst: O(n²)
- Space: O(1)
"""
function shellsort!(arr::Vector{T}) where T
    n = length(arr)
    gap = n ÷ 2

    while gap > 0
        for i in gap+1:n
            temp = arr[i]
            j = i

            while j > gap && arr[j - gap] > temp
                arr[j] = arr[j - gap]
                j -= gap
            end

            arr[j] = temp
        end

        gap ÷= 2
    end

    return arr
end

"""
    countingsort(arr::Vector{Int}, max_val::Int=maximum(arr))

Sort array of non-negative integers using counting sort.

# Time Complexity
- O(n + k) where k is the range
- Space: O(k)
"""
function countingsort(arr::Vector{Int}, max_val::Int=maximum(arr))
    if isempty(arr)
        return Int[]
    end

    min_val = minimum(arr)
    range_size = max_val - min_val + 1
    count = zeros(Int, range_size)
    output = similar(arr)

    # Count occurrences
    for num in arr
        count[num - min_val + 1] += 1
    end

    # Cumulative count
    for i in 2:range_size
        count[i] += count[i - 1]
    end

    # Build output array
    for i in length(arr):-1:1
        idx = arr[i] - min_val + 1
        output[count[idx]] = arr[i]
        count[idx] -= 1
    end

    return output
end

"""
    radixsort(arr::Vector{Int})

Sort array of non-negative integers using radix sort.

# Time Complexity
- O(nk) where k is the number of digits
- Space: O(n)
"""
function radixsort(arr::Vector{Int})
    if isempty(arr)
        return Int[]
    end

    result = copy(arr)
    max_num = maximum(arr)
    exp = 1

    while max_num ÷ exp > 0
        counting_sort_by_digit!(result, exp)
        exp *= 10
    end

    return result
end

function counting_sort_by_digit!(arr::Vector{Int}, exp::Int)
    n = length(arr)
    output = similar(arr)
    count = zeros(Int, 10)

    # Count occurrences of each digit
    for num in arr
        digit = (num ÷ exp) % 10
        count[digit + 1] += 1
    end

    # Cumulative count
    for i in 2:10
        count[i] += count[i - 1]
    end

    # Build output array
    for i in n:-1:1
        digit = (arr[i] ÷ exp) % 10
        output[count[digit + 1]] = arr[i]
        count[digit + 1] -= 1
    end

    # Copy back
    copyto!(arr, output)
end

"""
    bucketsort(arr::Vector{Float64}, n_buckets::Int=10)

Sort array of floats in range [0, 1) using bucket sort.

# Time Complexity
- Average: O(n + k)
- Worst: O(n²)
- Space: O(n)
"""
function bucketsort(arr::Vector{Float64}, n_buckets::Int=10)
    if isempty(arr)
        return Float64[]
    end

    # Initialize buckets
    buckets = [Float64[] for _ in 1:n_buckets]

    # Distribute elements into buckets
    for num in arr
        # Scale number to bucket range
        idx = min(Int(floor(num * n_buckets)) + 1, n_buckets)
        push!(buckets[idx], num)
    end

    # Sort individual buckets and concatenate
    result = Float64[]
    for bucket in buckets
        if !isempty(bucket)
            sort!(bucket)
            append!(result, bucket)
        end
    end

    return result
end

"""
    timsort!(arr::Vector{T}) where T

Sort array using TimSort algorithm (hybrid stable sort).

# Time Complexity
- Best: O(n)
- Average/Worst: O(n log n)
- Space: O(n)
"""
function timsort!(arr::Vector{T}) where T
    MIN_MERGE = 32
    n = length(arr)

    # Sort individual runs using insertion sort
    for start in 1:MIN_MERGE:n
        finish = min(start + MIN_MERGE - 1, n)
        insertionsort!(arr, start, finish)
    end

    # Start merging from MIN_MERGE
    size = MIN_MERGE
    while size < n
        for start in 1:2*size:n
            mid = start + size - 1
            finish = min(start + 2*size - 1, n)

            if mid < finish
                # Create auxiliary array for merge
                aux = similar(arr, finish - start + 1)
                merge_tim!(arr, aux, start, mid, finish)
            end
        end
        size *= 2
    end

    return arr
end

function merge_tim!(arr::Vector{T}, aux::Vector{T}, lo::Int, mid::Int, hi::Int) where T
    # Copy to auxiliary array
    k = 1
    for idx in lo:hi
        aux[k] = arr[idx]
        k += 1
    end

    i = 1
    j = mid - lo + 2

    for idx in lo:hi
        if i > mid - lo + 1
            arr[idx] = aux[j]
            j += 1
        elseif j > hi - lo + 1
            arr[idx] = aux[i]
            i += 1
        elseif aux[j] < aux[i]
            arr[idx] = aux[j]
            j += 1
        else
            arr[idx] = aux[i]
            i += 1
        end
    end
end

"""
    introsort!(arr::Vector{T}) where T

Sort array using IntroSort algorithm (hybrid of quicksort, heapsort, and insertion sort).

# Time Complexity
- All cases: O(n log n)
- Space: O(log n)
"""
function introsort!(arr::Vector{T}) where T
    max_depth = 2 * Int(floor(log2(length(arr))))
    introsort_helper!(arr, 1, length(arr), max_depth)
    return arr
end

function introsort_helper!(arr::Vector{T}, lo::Int, hi::Int, max_depth::Int) where T
    if hi - lo < 16
        insertionsort!(arr, lo, hi)
    elseif max_depth == 0
        heapsort_range!(arr, lo, hi)
    else
        p = partition!(arr, lo, hi)
        introsort_helper!(arr, lo, p - 1, max_depth - 1)
        introsort_helper!(arr, p + 1, hi, max_depth - 1)
    end
end

function heapsort_range!(arr::Vector{T}, lo::Int, hi::Int) where T
    # Build max heap for the range
    n = hi - lo + 1
    offset = lo - 1

    for i in n÷2:-1:1
        heapify_range!(arr, n, i, offset)
    end

    # Extract elements
    for i in n:-1:2
        arr[offset + 1], arr[offset + i] = arr[offset + i], arr[offset + 1]
        heapify_range!(arr, i - 1, 1, offset)
    end
end

function heapify_range!(arr::Vector{T}, n::Int, i::Int, offset::Int) where T
    largest = i
    left = 2 * i
    right = 2 * i + 1

    if left <= n && arr[offset + left] > arr[offset + largest]
        largest = left
    end

    if right <= n && arr[offset + right] > arr[offset + largest]
        largest = right
    end

    if largest != i
        arr[offset + i], arr[offset + largest] = arr[offset + largest], arr[offset + i]
        heapify_range!(arr, n, largest, offset)
    end
end

"""
    quickselect(arr::Vector{T}, k::Int) where T

Find the k-th smallest element in the array (1-indexed).

# Time Complexity
- Average: O(n)
- Worst: O(n²)
- Space: O(1)
"""
function quickselect(arr::Vector{T}, k::Int) where T
    if k < 1 || k > length(arr)
        throw(ArgumentError("k must be between 1 and length(arr)"))
    end

    arr_copy = copy(arr)
    return quickselect!(arr_copy, k, 1, length(arr_copy))
end

function quickselect!(arr::Vector{T}, k::Int, lo::Int, hi::Int) where T
    if lo == hi
        return arr[lo]
    end

    p = partition!(arr, lo, hi)

    if k == p
        return arr[k]
    elseif k < p
        return quickselect!(arr, k, lo, p - 1)
    else
        return quickselect!(arr, k, p + 1, hi)
    end
end

"""
    parallel_quicksort(arr::Vector{T}, threshold::Int=10000) where T

Sort array using parallel quicksort for large arrays.

# Time Complexity
- Average: O(n log n) with parallelism
- Space: O(log n)
"""
function parallel_quicksort(arr::Vector{T}, threshold::Int=10000) where T
    if length(arr) <= threshold
        return quicksort(arr)
    end

    result = copy(arr)
    parallel_quicksort!(result, 1, length(result), threshold)
    return result
end

function parallel_quicksort!(arr::Vector{T}, lo::Int, hi::Int, threshold::Int) where T
    if hi - lo < threshold
        quicksort!(arr, lo, hi)
        return
    end

    p = partition!(arr, lo, hi)

    # Parallel execution of left and right parts
    @sync begin
        Threads.@spawn parallel_quicksort!(arr, lo, p - 1, threshold)
        Threads.@spawn parallel_quicksort!(arr, p + 1, hi, threshold)
    end
end

end # module Sorting

File: javascript/src/sorting.js
/**
 * @module sorting
 * @description Sorting algorithms implementation
 */

/**
 * Swap two elements in an array
 * @private
 */
function swap(arr, i, j) {
    [arr[i], arr[j]] = [arr[j], arr[i]];
}

/**
 * Default comparison function
 * @private
 */
function defaultCompare(a, b) {
    if (a < b) return -1;
    if (a > b) return 1;
    return 0;
}

/**
 * Bubble Sort - O(n²) time, O(1) space
 * @param {Array} arr - Array to sort
 * @param {Function} [compareFunc] - Comparison function
 * @returns {Array} Sorted array
 */
export function bubbleSort(arr, compareFunc = defaultCompare) {
    const n = arr.length;
    const result = [...arr];

    for (let i = 0; i < n - 1; i++) {
        let swapped = false;
        for (let j = 0; j < n - i - 1; j++) {
            if (compareFunc(result[j], result[j + 1]) > 0) {
                swap(result, j, j + 1);
                swapped = true;
            }
        }
        if (!swapped) break;
    }

    return result;
}

/**
 * Insertion Sort - O(n²) time, O(1) space
 * @param {Array} arr - Array to sort
 * @param {Function} [compareFunc] - Comparison function
 * @returns {Array} Sorted array
 */
export function insertionSort(arr, compareFunc = defaultCompare) {
    const result = [...arr];
    const n = result.length;

    for (let i = 1; i < n; i++) {
        const key = result[i];
        let j = i - 1;

        while (j >= 0 && compareFunc(result[j], key) > 0) {
            result[j + 1] = result[j];
            j--;
        }
        result[j + 1] = key;
    }

    return result;
}

/**
 * Selection Sort - O(n²) time, O(1) space
 * @param {Array} arr - Array to sort
 * @param {Function} [compareFunc] - Comparison function
 * @returns {Array} Sorted array
 */
export function selectionSort(arr, compareFunc = defaultCompare) {
    const result = [...arr];
    const n = result.length;

    for (let i = 0; i < n - 1; i++) {
        let minIdx = i;
        for (let j = i + 1; j < n; j++) {
            if (compareFunc(result[j], result[minIdx]) < 0) {
                minIdx = j;
            }
        }
        if (minIdx !== i) {
            swap(result, i, minIdx);
        }
    }

    return result;
}

/**
 * Merge Sort - O(n log n) time, O(n) space
 * @param {Array} arr - Array to sort
 * @param {Function} [compareFunc] - Comparison function
 * @returns {Array} Sorted array
 */
export function mergeSort(arr, compareFunc = defaultCompare) {
    if (arr.length <= 1) return [...arr];

    const mid = Math.floor(arr.length / 2);
    const left = mergeSort(arr.slice(0, mid), compareFunc);
    const right = mergeSort(arr.slice(mid), compareFunc);

    return merge(left, right, compareFunc);
}

function merge(left, right, compareFunc) {
    const result = [];
    let i = 0, j = 0;

    while (i < left.length && j < right.length) {
        if (compareFunc(left[i], right[j]) <= 0) {
            result.push(left[i++]);
        } else {
            result.push(right[j++]);
        }
    }

    return result.concat(left.slice(i)).concat(right.slice(j));
}

/**
 * Quick Sort - O(n log n) average, O(n²) worst time, O(log n) space
 * @param {Array} arr - Array to sort
 * @param {Function} [compareFunc] - Comparison function
 * @returns {Array} Sorted array
 */
export function quickSort(arr, compareFunc = defaultCompare) {
    if (arr.length <= 1) return [...arr];

    const result = [...arr];
    quickSortHelper(result, 0, result.length - 1, compareFunc);
    return result;
}

function quickSortHelper(arr, low, high, compareFunc) {
    if (low < high) {
        const pivotIdx = partition(arr, low, high, compareFunc);
        quickSortHelper(arr, low, pivotIdx - 1, compareFunc);
        quickSortHelper(arr, pivotIdx + 1, high, compareFunc);
    }
}

function partition(arr, low, high, compareFunc) {
    const pivot = arr[high];
    let i = low - 1;

    for (let j = low; j < high; j++) {
        if (compareFunc(arr[j], pivot) <= 0) {
            i++;
            swap(arr, i, j);
        }
    }

    swap(arr, i + 1, high);
    return i + 1;
}

/**
 * Heap Sort - O(n log n) time, O(1) space
 * @param {Array} arr - Array to sort
 * @param {Function} [compareFunc] - Comparison function
 * @returns {Array} Sorted array
 */
export function heapSort(arr, compareFunc = defaultCompare) {
    const result = [...arr];
    const n = result.length;

    // Build heap
    for (let i = Math.floor(n / 2) - 1; i >= 0; i--) {
        heapify(result, n, i, compareFunc);
    }

    // Extract elements from heap
    for (let i = n - 1; i > 0; i--) {
        swap(result, 0, i);
        heapify(result, i, 0, compareFunc);
    }

    return result;
}

function heapify(arr, n, i, compareFunc) {
    let largest = i;
    const left = 2 * i + 1;
    const right = 2 * i + 2;

    if (left < n && compareFunc(arr[left], arr[largest]) > 0) {
        largest = left;
    }

    if (right < n && compareFunc(arr[right], arr[largest]) > 0) {
        largest = right;
    }

    if (largest !== i) {
        swap(arr, i, largest);
        heapify(arr, n, largest, compareFunc);
    }
}

/**
 * Radix Sort - O(nk) time, O(n+k) space (for integers)
 * @param {Array<number>} arr - Array of non-negative integers
 * @returns {Array<number>} Sorted array
 */
export function radixSort(arr) {
    if (arr.length === 0) return [];

    const result = [...arr];
    const max = Math.max(...result);

    for (let exp = 1; Math.floor(max / exp) > 0; exp *= 10) {
        countingSortByDigit(result, exp);
    }

    return result;
}

function countingSortByDigit(arr, exp) {
    const n = arr.length;
    const output = new Array(n);
    const count = new Array(10).fill(0);

    // Count occurrences of each digit
    for (let i = 0; i < n; i++) {
        const digit = Math.floor(arr[i] / exp) % 10;
        count[digit]++;
    }

    // Calculate cumulative count
    for (let i = 1; i < 10; i++) {
        count[i] += count[i - 1];
    }

    // Build output array
    for (let i = n - 1; i >= 0; i--) {
        const digit = Math.floor(arr[i] / exp) % 10;
        output[count[digit] - 1] = arr[i];
        count[digit]--;
    }

    // Copy output to arr
    for (let i = 0; i < n; i++) {
        arr[i] = output[i];
    }
}

/**
 * Counting Sort - O(n+k) time, O(k) space
 * @param {Array<number>} arr - Array of non-negative integers
 * @param {number} [maxValue] - Maximum value in array
 * @returns {Array<number>} Sorted array
 */
export function countingSort(arr, maxValue = null) {
    if (arr.length === 0) return [];

    const max = maxValue ?? Math.max(...arr);
    const min = Math.min(...arr);
    const range = max - min + 1;
    const count = new Array(range).fill(0);
    const result = new Array(arr.length);

    // Count occurrences
    for (const num of arr) {
        count[num - min]++;
    }

    // Calculate cumulative count
    for (let i = 1; i < range; i++) {
        count[i] += count[i - 1];
    }

    // Build output array
    for (let i = arr.length - 1; i >= 0; i--) {
        result[count[arr[i] - min] - 1] = arr[i];
        count[arr[i] - min]--;
    }

    return result;
}

/**
 * Bucket Sort - O(n+k) average time, O(n) space
 * @param {Array<number>} arr - Array of numbers in range [0,1)
 * @param {number} [bucketSize] - Number of buckets
 * @returns {Array<number>} Sorted array
 */
export function bucketSort(arr, bucketSize = 10) {
    if (arr.length === 0) return [];

    // Find range
    const min = Math.min(...arr);
    const max = Math.max(...arr);
    const range = max - min;

    // Create buckets
    const bucketCount = Math.floor(range / bucketSize) + 1;
    const buckets = Array.from({ length: bucketCount }, () => []);

    // Distribute elements into buckets
    for (const num of arr) {
        const bucketIdx = Math.floor((num - min) / bucketSize);
        buckets[bucketIdx].push(num);
    }

    // Sort each bucket and concatenate
    const result = [];
    for (const bucket of buckets) {
        if (bucket.length > 0) {
            result.push(...insertionSort(bucket));
        }
    }

    return result;
}

/**
 * Tim Sort - Hybrid stable sort (merge + insertion)
 * O(n log n) time, O(n) space
 * @param {Array} arr - Array to sort
 * @param {Function} [compareFunc] - Comparison function
 * @returns {Array} Sorted array
 */
export function timSort(arr, compareFunc = defaultCompare) {
    const MIN_MERGE = 32;
    const n = arr.length;
    const result = [...arr];

    // Sort individual runs using insertion sort
    for (let i = 0; i < n; i += MIN_MERGE) {
        const end = Math.min(i + MIN_MERGE, n);
        insertionSortRange(result, i, end, compareFunc);
    }

    // Merge runs
    let size = MIN_MERGE;
    while (size < n) {
        for (let start = 0; start < n; start += size * 2) {
            const mid = start + size;
            const end = Math.min(start + size * 2, n);
            if (mid < end) {
                mergeRange(result, start, mid, end, compareFunc);
            }
        }
        size *= 2;
    }

    return result;
}

function insertionSortRange(arr, left, right, compareFunc) {
    for (let i = left + 1; i < right; i++) {
        const key = arr[i];
        let j = i - 1;
        while (j >= left && compareFunc(arr[j], key) > 0) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
}

function mergeRange(arr, left, mid, right, compareFunc) {
    const leftPart = arr.slice(left, mid);
    const rightPart = arr.slice(mid, right);
    let i = 0, j = 0, k = left;

    while (i < leftPart.length && j < rightPart.length) {
        if (compareFunc(leftPart[i], rightPart[j]) <= 0) {
            arr[k++] = leftPart[i++];
        } else {
            arr[k++] = rightPart[j++];
        }
    }

    while (i < leftPart.length) {
        arr[k++] = leftPart[i++];
    }

    while (j < rightPart.length) {
        arr[k++] = rightPart[j++];
    }
}

/**
 * Shell Sort - O(n log n) to O(n²) time, O(1) space
 * @param {Array} arr - Array to sort
 * @param {Function} [compareFunc] - Comparison function
 * @returns {Array} Sorted array
 */
export function shellSort(arr, compareFunc = defaultCompare) {
    const result = [...arr];
    const n = result.length;

    // Start with large gap, reduce by half each iteration
    for (let gap = Math.floor(n / 2); gap > 0; gap = Math.floor(gap / 2)) {
        for (let i = gap; i < n; i++) {
            const temp = result[i];
            let j = i;

            while (j >= gap && compareFunc(result[j - gap], temp) > 0) {
                result[j] = result[j - gap];
                j -= gap;
            }

            result[j] = temp;
        }
    }

    return result;
}

/**
 * Comb Sort - O(n²) worst case, better average than bubble sort
 * @param {Array} arr - Array to sort
 * @param {Function} [compareFunc] - Comparison function
 * @returns {Array} Sorted array
 */
export function combSort(arr, compareFunc = defaultCompare) {
    const result = [...arr];
    const n = result.length;
    let gap = n;
    const shrink = 1.3;
    let sorted = false;

    while (gap > 1 || !sorted) {
        gap = Math.floor(gap / shrink);
        if (gap < 1) gap = 1;

        sorted = true;
        for (let i = 0; i + gap < n; i++) {
            if (compareFunc(result[i], result[i + gap]) > 0) {
                swap(result, i, i + gap);
                sorted = false;
            }
        }
    }

    return result;
}

/**
 * Gnome Sort - O(n²) time, O(1) space
 * @param {Array} arr - Array to sort
 * @param {Function} [compareFunc] - Comparison function
 * @returns {Array} Sorted array
 */
export function gnomeSort(arr, compareFunc = defaultCompare) {
    const result = [...arr];
    const n = result.length;
    let i = 0;

    while (i < n) {
        if (i === 0 || compareFunc(result[i], result[i - 1]) >= 0) {
            i++;
        } else {
            swap(result, i, i - 1);
            i--;
        }
    }

    return result;
}

/**
 * Cocktail Sort (Bidirectional Bubble Sort) - O(n²) time, O(1) space
 * @param {Array} arr - Array to sort
 * @param {Function} [compareFunc] - Comparison function
 * @returns {Array} Sorted array
 */
export function cocktailSort(arr, compareFunc = defaultCompare) {
    const result = [...arr];
    const n = result.length;
    let left = 0;
    let right = n - 1;
    let swapped = true;

    while (swapped && left < right) {
        swapped = false;

        // Left to right
        for (let i = left; i < right; i++) {
            if (compareFunc(result[i], result[i + 1]) > 0) {
                swap(result, i, i + 1);
                swapped = true;
            }
        }
        right--;

        if (!swapped) break;

        swapped = false;

        // Right to left
        for (let i = right; i > left; i--) {
            if (compareFunc(result[i], result[i - 1]) < 0) {
                swap(result, i, i - 1);
                swapped = true;
            }
        }
        left++;
    }

    return result;
}

/**
 * Check if an array is sorted
 * @param {Array} arr - Array to check
 * @param {Function} [compareFunc] - Comparison function
 * @returns {boolean} True if sorted
 */
export function isSorted(arr, compareFunc = defaultCompare) {
    for (let i = 1; i < arr.length; i++) {
        if (compareFunc(arr[i - 1], arr[i]) > 0) {
            return false;
        }
    }
    return true;
}

/**
 * Shuffle an array using Fisher-Yates algorithm
 * @param {Array} arr - Array to shuffle
 * @returns {Array} Shuffled array
 */
export function shuffle(arr) {
    const result = [...arr];
    const n = result.length;

    for (let i = n - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        swap(result, i, j);
    }

    return result;
}

/**
 * Benchmark a sorting algorithm
 * @param {Function} sortFunc - Sorting function to benchmark
 * @param {Array} arr - Array to sort
 * @returns {Object} Benchmark results
 */
export function benchmarkSort(sortFunc, arr) {
    const startTime = performance.now();
    const sorted = sortFunc([...arr]);
    const endTime = performance.now();

    return {
        algorithm: sortFunc.name,
        inputSize: arr.length,
        timeMs: endTime - startTime,
        isSorted: isSorted(sorted)
    };
}

/**
 * Compare multiple sorting algorithms
 * @param {Array} algorithms - Array of sorting functions
 * @param {Array} arr - Test array
 * @returns {Array} Comparison results
 */
export function compareSorts(algorithms, arr) {
    return algorithms.map(algo => benchmarkSort(algo, arr))
        .sort((a, b) => a.timeMs - b.timeMs);
}

// Export all sorting algorithms
export default {
    bubbleSort,
    insertionSort,
    selectionSort,
    mergeSort,
    quickSort,
    heapSort,
    radixSort,
    countingSort,
    bucketSort,
    timSort,
    shellSort,
    combSort,
    gnomeSort,
    cocktailSort,
    isSorted,
    shuffle,
    benchmarkSort,
    compareSorts
};